base/shell/explorer/desktop.cpp

```diff
@@ -22,97 +22,163 @@
+        return E_FAIL;
     }
 
-public:
-    CDesktopThread() :
-        m_hEvent(NULL),
-        m_Tray(NULL)
-    {
+    m_Tray = pTray;
+    m_hThread = CreateThread(NULL, 0, s_DesktopThreadProc, (LPVOID)this, 0, NULL);
+
+    if (!m_hThread)
+    {   
+        CloseHandle(m_hInitEvent);
+        m_hInitEvent = NULL;
+
+        m_Tray = NULL;
+
+        return E_FAIL;
     }
 
-    HRESULT Initialize(IN OUT ITrayWindow *pTray)
+    Handles[0] = m_hThread;
+    Handles[1] = m_hInitEvent;
+
+    for (;;)
     {
-        HANDLE hThread;
-        HANDLE Handles[2];
+        DWORD WaitResult = MsgWaitForMultipleObjects(_countof(Handles), Handles, FALSE, INFINITE, QS_ALLEVENTS);
 
-        m_Tray = pTray;
+        if (WaitResult == WAIT_OBJECT_0 + _countof(Handles))
+        {
+            TrayProcessMessages(m_Tray);
+        }
+        else if (WaitResult != WAIT_FAILED && WaitResult != WAIT_OBJECT_0)
+        {
+            break;
+        }
+        else
+        {
+            CloseHandle(m_hThread);
+            m_hThread = NULL;
 
-        m_hEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
-        if (!m_hEvent)
-            return E_FAIL;
+            CloseHandle(m_hInitEvent);
+            m_hInitEvent = NULL;
+
+            m_Tray = NULL;
 
-        hThread = CreateThread(NULL, 0, s_DesktopThreadProc, (PVOID)this, 0, NULL);
-        if (!hThread)
-        {
-            CloseHandle(m_hEvent);
             return E_FAIL;
         }
+    }
+    
+    return S_OK;
+}
 
-        Handles[0] = hThread;
-        Handles[1] = m_hEvent;
-
-        for (;;)
+void CDesktopThread::Destroy()
+{
+    if (m_hThread)
+    {
+        DWORD WaitResult = WaitForSingleObject(m_hThread, 0);
+        
+        if (WaitResult == WAIT_TIMEOUT)
         {
-            DWORD WaitResult = MsgWaitForMultipleObjects(_countof(Handles), Handles, FALSE, INFINITE, QS_ALLEVENTS);
-            if (WaitResult == WAIT_OBJECT_0 + _countof(Handles))
-            {
-                TrayProcessMessages(m_Tray);
-            }
-            else if (WaitResult != WAIT_FAILED && WaitResult != WAIT_OBJECT_0)
-            {
-                break;
-            }
+            /* Send WM_QUIT message to the thread and wait for it to terminate */
+            PostThreadMessageW(GetThreadId(m_hThread), WM_QUIT, 0, 0);
+            WaitForSingleObject(m_hThread, INFINITE);
         }
 
-        CloseHandle(hThread);
-        CloseHandle(m_hEvent);
+        CloseHandle(m_hThread);
+        m_hThread = NULL;
+    }
 
-        return S_OK;
+    if (m_hInitEvent)
+    {
+        CloseHandle(m_hInitEvent);
+        m_hInitEvent = NULL;
     }
 
-    void Destroy()
+    m_Tray = NULL;
+}
+
+DWORD CDesktopThread::DesktopThreadProc()
+{
+    CComPtr<IShellDesktopTray> pSdt;
+    HANDLE hDesktop;
+    HRESULT hRet;
+
+    OleInitialize(NULL);
+
+    hRet = m_Tray->QueryInterface(IID_PPV_ARG(IShellDesktopTray, &pSdt));
+    if (!SUCCEEDED(hRet))
     {
-        return;
+        return 1;
```

Comment:
Shouldn't we call OleUninitialize in each of these failure return paths?

Reply:
Fixed.